cpp/include/cudf/table/experimental/row_operators.cuh

```diff
@@ -469,6 +475,187 @@ class self_comparator {
   std::shared_ptr<preprocessed_table> d_t;
 };
 
+enum class lhs_index_type : size_type {};
+enum class rhs_index_type : size_type {};
+
+struct make_lhs_index {
+  __device__ lhs_index_type operator()(size_type i) const { return static_cast<lhs_index_type>(i); }
+};
+
+struct make_rhs_index {
+  __device__ rhs_index_type operator()(size_type i) const { return static_cast<rhs_index_type>(i); }
+};
+
+auto inline make_lhs_index_counting_iterator(size_type start)
+{
+  return cudf::detail::make_counting_transform_iterator(start, make_lhs_index{});
+};
+
+auto inline make_rhs_index_counting_iterator(size_type start)
+{
+  return cudf::detail::make_counting_transform_iterator(start, make_rhs_index{});
+};
+
+template <typename Nullate>
+class two_table_device_row_comparator_adapter {
+  friend class two_table_comparator;
+
+ public:
+  /**
+   * @brief Checks whether the row at `lhs_index` in the `lhs` table compares
+   * lexicographically less than the row at `rhs_index` in the `rhs` table.
+   *
+   * @param lhs_index The index of the row in the `lhs` table to examine
+   * @param rhs_index The index of the row in the `rhs` table to examine
+   * @return `true` if row from the `lhs` table compares less than row in the `rhs` table
+   */
+  __device__ weak_ordering operator()(lhs_index_type const lhs_index,
+                                      rhs_index_type const rhs_index) const noexcept
+  {
+    return comp(static_cast<cudf::size_type>(lhs_index), static_cast<cudf::size_type>(rhs_index));
```

Comment:
```suggestion
    return comp(static_cast<cudf::size_type>(lhs_index), static_cast<cudf::size_type>(rhs_index)) ==
           weak_ordering::LESS;
```

Reply:
I'm thinking about this a little more. We want to use the weak ordering machinery if possible, and separate these layers.

Reply:
We are going to call `thrust::lower_bound` on the comparator. If we want to keep the `weak_ordering` result, we must have another separate layer for it. This layer must return bool.

Reply:
See: https://github.com/rapidsai/cudf/pull/10730/files#r872743767